File: framework/src/main/java/nostalgia/framework/ui/gamegallery/AppStoreAdapter.java
package nostalgia.framework.ui.gamegallery;

import android.content.Context;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.SectionIndexer;
import android.widget.TextView;

import com.blankj.utilcode.util.Utils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Set;

import nostalgia.framework.R;
import nostalgia.framework.utils.NLog;

public class AppStoreAdapter extends BaseAdapter implements SectionIndexer {

    private static final String[] names = Utils.getApp().getResources().getStringArray(R.array.gallery_page_tab_names);

    enum SORT_TYPES {
        SORT_BY_NAME_ALPHA {
            public String getTabName() {
                return names[0];
            }
        },
        SORT_BY_INSERT_DATE {
            public String getTabName() {
                return names[1];
            }
        },
        SORT_BY_MOST_PLAYED {
            public String getTabName() {
                return names[2];
            }
        },
        SORT_BY_LAST_PLAYED {
            public String getTabName() {
                return names[3];
            }
        },
        DOWNLOAD_BY_REMOTE {
            public String getTabName() {
                return names[4];
            }
        };

        public abstract String getTabName();
    }

    private HashMap<Character, Integer> alphaIndexer = new HashMap<>();
    private String filter = "";
    private Character[] sections;
    private LayoutInflater inflater;
    private Context context;
    private int mainColor;
    private ArrayList<GameDescription> games = new ArrayList<>();
    private ArrayList<RowItem> filterGames = new ArrayList<>();
    private int sumRuns = 0;
    private SORT_TYPES sortType = SORT_TYPES.DOWNLOAD_BY_REMOTE;

    private Comparator<GameDescription> nameComparator = (lhs, rhs) ->
            lhs.getSortName().compareTo(rhs.getSortName());

    private Comparator<GameDescription> insertDateComparator = (lhs, rhs) ->
            (int) (-lhs.inserTime + rhs.inserTime);

    private Comparator<GameDescription> lastPlayedDateComparator = (lhs, rhs) -> {
        long dif = lhs.lastGameTime - rhs.lastGameTime;
        if (dif == 0) {
            return 0;
        } else if (dif < 0) {
            return 1;
        } else {
            return -1;
        }
    };
    private Comparator<GameDescription> playedCountComparator = (lhs, rhs) ->
            -lhs.runCount + rhs.runCount;

    public AppStoreAdapter(Context context) {
        this.context = context;
        inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        mainColor = context.getResources().getColor(R.color.main_color);
        GameDescription gameDescription = new GameDescription();
        gameDescription.name = "坦克";
        games.add(gameDescription);

        for (GameDescription game : games) {
            sumRuns = game.runCount > sumRuns ? game.runCount : sumRuns;
            String name = game.getCleanName().toLowerCase();
            RowItem item = new RowItem();
            item.game = game;
            item.firstLetter = name.charAt(0);
            filterGames.add(item);
        }
    }

    @Override
    public int getCount() {
        return filterGames.size();
    }

    @Override
    public Object getItem(int position) {
        return filterGames.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        RowItem item = filterGames.get(position);
        if (convertView == null) {
            convertView = inflater.inflate(R.layout.row_game_list, null);
        }
        GameDescription game = item.game;
        TextView name = convertView.findViewById(R.id.row_game_item_name);
        ImageView arrowIcon = convertView.findViewById(R.id.game_item_arrow);
        ImageView bck = convertView.findViewById(R.id.game_item_bck);
        ProgressBar runIndicator = convertView.findViewById(R.id.row_game_item_progressBar);
        runIndicator.setMax(sumRuns);
        name.setText(game.getCleanName());
        arrowIcon.setImageResource(R.drawable.ic_next_arrow);
        arrowIcon.clearAnimation();
        name.setTextColor(mainColor);
        name.setGravity(Gravity.CENTER_VERTICAL);
        bck.setImageResource(R.drawable.game_item_small_bck);
        return convertView;
    }

    @Override
    public int getPositionForSection(int section) {
        try {
            Character ch = Character.toLowerCase(sections[section]);
            Integer pos = alphaIndexer.get(ch);
            if (pos == null) {
                return 0;
            } else {
                return pos;
            }
        } catch (ArrayIndexOutOfBoundsException e) {
            return 0;
        }
    }

    @Override
    public int getSectionForPosition(int position) {
        RowItem item = (RowItem) getItem(position);
        char ch = Character.toUpperCase(item.firstLetter);
        for (int i = 0; i < sections.length; i++) {
            Character ch1 = sections[i];
            if (ch1.equals(ch)) {
                return i;
            }
        }
        return 1;
    }

    @Override
    public Object[] getSections() {
        Set<Character> keyset = alphaIndexer.keySet();
        sections = new Character[keyset.size()];
        keyset.toArray(sections);
        Arrays.sort(sections, Character::compareTo);
        for (int i = 0; i < sections.length; i++)
            sections[i] = Character.toUpperCase(sections[i]);
        return sections;
    }

    @Override
    public void notifyDataSetChanged() {

    }

    public class RowItem {
        GameDescription game;
        char firstLetter;
    }

}

File: gui/src/main/java/com/woohyman/gui/ui/gamegallery/adapter/AppStoreAdapter.kt
package com.woohyman.gui.ui.gamegallery.adapter

import android.content.Context
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.SectionIndexer
import android.widget.TextView
import com.blankj.utilcode.util.Utils
import com.woohyman.gui.R
import com.woohyman.keyboard.data.database.GameDescription
import com.woohyman.keyboard.data.entity.RowItem
import java.util.Arrays
import java.util.Locale

class AppStoreAdapter(private val context: Context) : BaseAdapter(), SectionIndexer {
    internal enum class SORT_TYPES {
        SORT_BY_NAME_ALPHA {
            override val tabName: String
                get() = names[0]
        },
        SORT_BY_INSERT_DATE {
            override val tabName: String
                get() = names[1]
        },
        SORT_BY_MOST_PLAYED {
            override val tabName: String
                get() = names[2]
        },
        SORT_BY_LAST_PLAYED {
            override val tabName: String
                get() = names[3]
        },
        DOWNLOAD_BY_REMOTE {
            override val tabName: String
                get() = names[4]
        };

        abstract val tabName: String
    }

    private val alphaIndexer = HashMap<Char, Int>()
    private val filter = ""
    private var sections: Array<Char?> = emptyArray()
    private val inflater: LayoutInflater
    private val mainColor: Int
    private val games = ArrayList<GameDescription>()
    private val filterGames = ArrayList<RowItem>()
    private var sumRuns = 0
    private val sortType = SORT_TYPES.DOWNLOAD_BY_REMOTE
    private val nameComparator =
        java.util.Comparator { lhs: GameDescription, rhs: GameDescription ->
            lhs.sortName.compareTo(rhs.sortName)
        }
    private val insertDateComparator =
        java.util.Comparator { lhs: GameDescription, rhs: GameDescription -> (-lhs.inserTime + rhs.inserTime).toInt() }
    private val lastPlayedDateComparator =
        java.util.Comparator { lhs: GameDescription, rhs: GameDescription ->
            val dif = lhs.lastGameTime - rhs.lastGameTime
            if (dif == 0L) {
                return@Comparator 0
            } else if (dif < 0) {
                return@Comparator 1
            } else {
                return@Comparator -1
            }
        }
    private val playedCountComparator =
        java.util.Comparator { lhs: GameDescription, rhs: GameDescription -> -lhs.runCount + rhs.runCount }

    init {
        inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        mainColor = context.resources.getColor(R.color.main_color)
        val gameDescription = GameDescription()
        gameDescription.url =
            "https://gitee.com/popvan/nes-repo/raw/master/roms/Super%20Mario%20Bros.%203.nes"
        gameDescription.name = "超级马里奥兄弟3"
        games.add(gameDescription)
        for (game in games) {
            sumRuns = if (game.runCount > sumRuns) game.runCount else sumRuns
            val name = game.cleanName.lowercase(Locale.getDefault())
            val item = RowItem()
            item.game = game
            item.firstLetter = name[0]
            filterGames.add(item)
        }
    }

    override fun getCount(): Int {
        return filterGames.size
    }

    override fun getItem(position: Int): Any {
        return filterGames[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return convertView ?: inflater.inflate(R.layout.row_game_list, null).also {
            val (game) = filterGames[position]
            val name = it.findViewById<TextView>(R.id.row_game_item_name)
            val arrowIcon = it.findViewById<ImageView>(R.id.game_item_arrow)
            val bck = it.findViewById<ImageView>(R.id.game_item_bck)
            name.text = game!!.cleanName
            arrowIcon.setImageResource(R.drawable.ic_next_arrow)
            arrowIcon.clearAnimation()
            name.setTextColor(mainColor)
            name.gravity = Gravity.CENTER_VERTICAL
            bck.setImageResource(R.drawable.game_item_small_bck)
        }
    }

    override fun getPositionForSection(section: Int): Int {
        return try {
            val ch = sections[section]?.lowercaseChar()
            val pos = alphaIndexer[ch]
            pos ?: 0
        } catch (e: ArrayIndexOutOfBoundsException) {
            0
        }
    }

    override fun getSectionForPosition(position: Int): Int {
        val (_, firstLetter) = getItem(position) as RowItem
        val ch = firstLetter.uppercaseChar()
        for (i in sections.indices) {
            val ch1 = sections[i]
            if (ch1 == ch) {
                return i
            }
        }
        return 1
    }

    override fun getSections(): Array<Char?> {
        val keyset: Set<Char> = alphaIndexer.keys
        sections = arrayOfNulls(keyset.size)
//        keyset.toArray<Char>(sections)
        Arrays.sort(sections) { obj: Char?, anotherCharacter: Char? ->
            obj!!.compareTo(
                anotherCharacter!!
            )
        }
        for (i in sections.indices) sections[i] = sections[i]?.uppercaseChar()
        return sections
    }

    override fun notifyDataSetChanged() {}

    companion object {
        private val names = Utils.getApp().resources.getStringArray(R.array.gallery_page_tab_names)
    }
}